File: CarHomeWidget/src/com/anod/car/home/prefs/AllAppsActivity.java
package com.anod.car.home.prefs;

import java.util.ArrayList;

import android.app.ListActivity;
import android.content.ComponentName;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;

import com.anod.car.home.CarWidgetApplication;
import com.anod.car.home.R;
import com.anod.car.home.appscache.AppsCacheAdapter;
import com.anod.car.home.appscache.AppsCacheAsyncTask;
import com.anod.car.home.appscache.AppsCacheAsyncTask.Callback;
import com.anod.car.home.model.AppsListCache;
import com.anod.car.home.model.AppsListCache.CacheEntry;

public class AllAppsActivity extends ListActivity implements OnItemClickListener, Callback {
	private AppsListCache mAppsList;
	private ArrayList<CacheEntry> mAppsCacheEntries;
	private PackageManager mPackageManager;
	
    @Override
    protected void onCreate(Bundle savedInstanceState){
    	requestWindowFeature(Window.FEATURE_NO_TITLE);
        super.onCreate(savedInstanceState);
        getListView().setOnItemClickListener(this);
        setVisible(false);
        
        CarWidgetApplication app = (CarWidgetApplication)this.getApplicationContext();
        app.initAppListCache();
        
        mAppsList = app.getAppListCache();
        mAppsCacheEntries = (mAppsList == null) ? null : mAppsList.getCacheEntries();
        mPackageManager = getPackageManager();

        if (mAppsCacheEntries == null || mAppsCacheEntries.size() == 0) {
        	new AppsCacheAsyncTask(this, mAppsList, this).execute(0);
        } else {
        	showList();
        } 
    }
    
    private void showList(){
        // Now create a new list adapter bound to the cursor.
        // SimpleListAdapter is designed for binding to a Cursor.
        AppsCacheAdapter adapter = new AppsCacheAdapter(this, R.layout.all_apps_row, mAppsCacheEntries, mAppsList);
        // Bind to our new adapter.
        setListAdapter(adapter);
        try {
        	setVisible(true);
        } catch(Exception e) {
        	Log.d("CarHomeWidget", "Cannot set list visible");
        }
    }
    
    final private Intent getActivityIntent(ComponentName className) {
        Intent intent = new Intent(Intent.ACTION_MAIN);
        intent.addCategory(Intent.CATEGORY_LAUNCHER);
        intent.setComponent(className);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED);
        return intent;
	}

	@Override
	public void onItemClick(AdapterView<?> parent, View view,
	        int position, long id) {
		CacheEntry entry = mAppsCacheEntries.get(position);

        Intent intent = getActivityIntent(entry.componentName);
        setResult(RESULT_OK, intent);
        finish();
	}

	@Override
	public void onIntentFilterInit(Intent intent) {
		intent.addCategory(Intent.CATEGORY_LAUNCHER);
	}

	@Override
	public void onResult(ArrayList<CacheEntry> cacheEntries) {
		mAppsCacheEntries = cacheEntries;
		showList();
	}

}

File: CarHomeWidget/src/main/java/com/anod/car/home/prefs/AllAppsActivity.kt
package com.anod.car.home.prefs

import android.app.Activity
import com.anod.car.home.app.App
import com.anod.car.home.appscache.AppsCacheActivity
import com.anod.car.home.model.AppsList

import android.content.ComponentName
import android.content.Intent
import android.os.Bundle

class AllAppsActivity : AppsCacheActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel.appsList = App.provide(this).appListCache
    }

    override fun onEntryClick(position: Int, entry: AppsList.Entry) {
        val intent = createActivityIntent(entry.componentName)
        setResult(Activity.RESULT_OK, intent)
        finish()
    }

    private fun createActivityIntent(componentName: ComponentName?): Intent {
        val intent = Intent(Intent.ACTION_MAIN)
        intent.addCategory(Intent.CATEGORY_LAUNCHER)
        intent.component = componentName
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED
        return intent
    }

    override fun createQueryIntent(): Intent {
        val intent = Intent()
        intent.action = Intent.ACTION_MAIN
        intent.addCategory(Intent.CATEGORY_LAUNCHER)
        return intent
    }

}
